kubelet external componentconfig

mikedanese opened the pull request.

mikedanese commented:
cc @kubernetes/goog-node this is ready for review. Thanks!

mikedanese commented:
@zmerlynn @roberthbailey to review salt changes

mikedanese commented:
@timstclair this is ready for a review now

timstclair commented:
First review pass complete. I left a lot of comments on the API, but I wasn't sure if you were leaving that for a later revision (I see the internal API already went in and this just reflects those values, but it's a lot easier to change the internal version). I didn't look at the salt too closely, but I might not be the right person for a review of that anyway.

vishh commented on docs/admin/kubelet.md:
```diff
@@ -80,7 +81,7 @@ kubelet
       --containerized[=false]: Experimental support for running kubelet in a container.  Intended for testing. [default=false]
       --cpu-cfs-quota[=true]: Enable CPU CFS quota enforcement for containers that specify CPU limits
-      --docker-endpoint="": If non-empty, use this for the docker endpoint to communicate with
+      --docker-endpoint="unix:///var/run/docker.sock": If non-empty, use this for the docker endpoint to communicate with
```
Given that we are defaulting now, should we include the phrase ` If non-empty`? Even the defaults are explicit right?

vishh commented on pkg/apis/componentconfig/v1alpha1/types.go:
```diff
@@ -74,3 +74,219 @@ const (
 	ProxyModeIPTables  ProxyMode = "iptables"
 )
+
+type KubeletConfiguration struct {
```
Question: How do we know which fields were optional in the flag based config? Are we relying on flag description?

mikedanese commented:
So we don't lose them, tim's comments are here: https://github.com/mikedanese/kubernetes/commit/d2b4974778ed92b673812fc1c0a30054f72dc501

Working on addressing them now.

mikedanese commented:
> > +20,599 −15,889

> Labelling this PR as size/XS

I think i broke @k8s-merge-robot

vishh commented:
@mikedanese: :clap:

derekwaynecarr commented on cmd/kubelet/app/options/options.go:
```diff
@@ -60,74 +51,16 @@ type KubeletServer struct {
-			CertDirectory:               "/var/run/kubernetes",
-			CgroupRoot:                  "",
-			ConfigureCBR0:               false,
-			ContainerRuntime:            "docker",
```
where did all the defaults get moved?

looking at the generated docs they are lost... for example, container runtime in the generated docs shows "" when it was "docker"

mikedanese replied:
yes, i think I broke something in a rebase. They have moved to pkg/apis/componentconfig/v1alpha1/defaults.go.

Assuming the weren't broken, they would show up in the .md files. Equivalent PR for kube proxy https://github.com/kubernetes/kubernetes/pull/20542/files

derekwaynecarr commented on cmd/kubelet/app/options/options.go:
```diff
@@ -60,74 +51,16 @@ type KubeletServer struct {
 
 // NewKubeletServer will create a new KubeletServer with default values.
 func NewKubeletServer() *KubeletServer {
+	config := componentconfig.KubeletConfiguration{}
```
@liggitt - fyi, still reviewing this, but it looks like defaulting has been moved here....

liggitt replied:
as long as the returned `KubeletServer` still contains a defaulted config, it doesn't impact us

derekwaynecarr commented on docs/admin/kubelet.md:
```diff
@@ -77,12 +77,12 @@ kubelet
-      --configure-hairpin-mode[=true]: If true, kubelet will set the hairpin mode flag on container interfaces. This allows endpoints of a Service to loadbalance back to themselves if they should try to access their own Service.
-      --container-runtime="docker": The container runtime to use. Possible values: 'docker', 'rkt'. Default: 'docker'.
+      --configure-hairpin-mode[=false]: If true, kubelet will set the hairpin mode flag on container interfaces. This allows endpoints of a Service to loadbalance back to themselves if they should try to access their own Service.
+      --container-runtime="": The container runtime to use. Possible values: 'docker', 'rkt'. Default: 'docker'.
```
This is the one that stands out as confusing me, so maybe this needs regenerating?

I see the defaults.go sets the default to "docker" down below, but it appears to be lost here...

mikedanese commented:
@derekwaynecarr fixed

derekwaynecarr commented:
Cool.  I'm excited to use this.

On Thursday, February 11, 2016, Mike Danese <notifications@github.com>
wrote:

> @derekwaynecarr <https://github.com/derekwaynecarr> fixed
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/19837#issuecomment-183119765>
> .
>

derekwaynecarr commented:
Do we plan to support reading config from file source in 1.2?  It would
help prototype some stuff that is harder to represent in flags (pod
eviction that I am prototyping)

On Thursday, February 11, 2016, Derek Carr <decarr@redhat.com> wrote:

> Cool.  I'm excited to use this.
>
> On Thursday, February 11, 2016, Mike Danese <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
>> @derekwaynecarr <https://github.com/derekwaynecarr> fixed
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/19837#issuecomment-183119765>
>> .
>>
>

mikedanese commented:
Yes hoping to get the file based config in for 1.2, I'm planning on submitting this https://github.com/kubernetes/kubernetes/commit/aa31a4cb54978801563c8bb73dd46d46a04fec63 after this PR is merged

mikedanese commented:
@timothysc I think I addresssed all comments, can you do another pass?

timstclair commented:
Ack (you mean @timstclair) - I'll take a look later today.

mikedanese commented:
Yes sorry, wrong tim. Thanks. Build failed in contrib/mesos, will fix and rebase (for the 9th time :). I want to stop rebasing but I'm also kind of proud. This might be a new record.

liggitt commented on pkg/apis/componentconfig/v1alpha1/types.go:
```diff
@@ -127,3 +127,234 @@ type LeaderElectionConfiguration struct {
+	unversioned.TypeMeta
+
+	// config is the path to the config file or directory of files
+	Config string `json:"config"`
```
Is this relative or absolute? If relative, is it resolved relative to the location of the file containing this config, or to the cwd of the kubelet process, or something else?

liggitt replied:
Same question goes for all file/dir/path fields

liggitt commented on pkg/apis/componentconfig/v1alpha1/types.go:
```diff
@@ -127,3 +127,234 @@ type LeaderElectionConfiguration struct {
+	// Defaults to false.
+	AllowPrivileged *bool `json:"allowPrivileged"`
+	// hostNetworkSources is a comma-separated list of sources from which the
+	// Kubelet allows pods to use of host network. Defaults to "*".
```
Defaults to * if nil, but not if empty, right?

vishh replied:
ping @mtaufen

mtaufen replied:
Yes, the zero value for a slice is `nil`, and we perform the following check in `pkg/apis/componentconfig/v1alpha1/defaults.go`:

```
if obj.HostNetworkSources == nil {
	obj.HostNetworkSources = []string{kubetypes.AllSource}
}
```

`kubetypes.AllSource` is `*`

liggitt commented on pkg/apis/componentconfig/v1alpha1/types.go:
```diff
@@ -127,3 +127,234 @@ type LeaderElectionConfiguration struct {
+	// eventBurst is the maximum size of a bursty event records, temporarily
+	// allows event records to burst to this number, while still not exceeding
+	// event-qps. Only used if eventQps > 0
+	EventBurst int64 `json:"eventBurst"`
```
Do we need int64 instead of int32? See API conventions about issues with 64-bit ints as a cautionary tale :)

liggitt replied:
(it may be needed, which is fine, but care should be taken in validation/conversion to avoid ints > 52 bits if we're serializing to JSON)

mtaufen replied:
On further inspection, there are several `int64` fields in the external type in this PR that will be down-converted to `int32` in the internal type. Wouldn't it make sense to just make these fields `int32` in the external type as well?

Fields which are `int32` internally but `int64` externally:
- `RegistryBurst`
- `EventBurst`
- `MaxPerPodContainerCount`
- `LowDiskSpaceThresholdMB`

timstclair replied:
Yes.

timstclair commented on pkg/apis/componentconfig/v1alpha1/defaults.go:
```diff
@@ -84,16 +86,186 @@ func addDefaultingFuncs(scheme *runtime.Scheme) {
+			if obj.EventBurst == 0 {
+				obj.EventBurst = 10
+			}
+			if obj.EventRecordQPS == 0 {
```
This is not consistent with the documentation of this field:
> If 0, there is no limit enforced.

This is why we need pointers for primitives...

timstclair commented on pkg/apis/componentconfig/v1alpha1/defaults.go:
```diff
@@ -84,16 +86,186 @@ func addDefaultingFuncs(scheme *runtime.Scheme) {
+				obj.RegistryBurst = 10
+			}
+			if obj.RegistryPullQPS == 0 {
+				obj.RegistryPullQPS = 5.0
```
Ditto re: docs say 0 for unlimited.

timstclair commented on pkg/apis/componentconfig/v1alpha1/types.go:
```diff
@@ -127,3 +127,234 @@ type LeaderElectionConfiguration struct {
+	// readOnlyPort is the read-only port for the Kubelet to serve on with
+	// no authentication/authorization (set to 0 to disable)
+	ReadOnlyPort int32 `json:"readOnlyPort"`
+	// tLSCertFile is the file containing x509 Certificate for HTTPS.  (CA cert,
```
(nit) capitalization of `tLSCertFile`

Actually, I think all the comments should be capitalized to match the fields...

mikedanese replied:
The comments should match the json field tags in API types.

vishh replied:
@mtaufen this comments needs to be addressed.

mtaufen replied:
I take it we mandate that json field names start with a lowercase letter? IMO I rank the following options in this order (most to least "pretty"): `TLSPrivateKeyFile`, then `tlsPrivateKeyFile`, then `tLSPrivateKeyFile`. Having the mixed-case letters in this initialism is ugly and odd to read.

timstclair replied:
Convention is for consistent capitalization of acronyms, so in this case the correct capitalization is `TLSPrivateKeyFile` for the go field name and `tlsPrivateKeyFile` for the JSON name. See https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#naming-conventions

mtaufen replied:
Ok, convention trumps my personal preference :). The same it shall remain.

timstclair replied:
Well, the json name and the comment are still wrong according to the convention.

mtaufen replied:
I'm seeing this in the convention:
`Other than capitalization of the initial letter, the two should almost always match.`

mtaufen replied:
Ah, nvm, I read `tlsPrivateKeyFile` in your comment as `tLSPrivateKeyFile`. Silly brain. I will update it to `tlsPrivateKeyFile`.

mikedanese committed:
componentconfig: add external kubelet configuration type

mikedanese commented:
@madhusudancs @liggitt @wojtek-t what could be wrong with my build environment that is creating this huge diff?

liggitt commented:
old codecgen, perhaps? not sure, other than the usual refetch, rebase, make clean, hack/update-codecgen.sh

madhusudancs commented:
@liggitt @mikedanese and I spent an hour debugging this. We did all these, plus we copied his codecgen_binary to my machine and ran hack/update-codecgen.sh with his codecgen_binary. And my machine generates the results as expected and his machine doesn't. What is even worse is that, it (codecgen on Mike's computer) doesn't seem to obey your vars reset changes here - https://github.com/ugorji/go/commit/b689e5578429a414e5e40df5fccc3ae03de4f466

mikedanese commented:
I even tried to isolate the run in a docker image with no luck:
```sh
#! /bin/bash

set -o errexit
set -o nounset
set -o pipefail

WORKDIR="/go/src/k8s.io/kubernetes"

cat <<EOF | docker build -t test -
FROM gcr.io/google_containers/kube-cross:v1.4.2-1
RUN apt-get update \
  && apt-get install -y vim
WORKDIR ${WORKDIR}
EOF

docker run \
  -ti \
  -v "${HOME}${WORKDIR}:${WORKDIR}" \
  test bash
```

From a fresh master, I was still generating diffs. I will keep investigating.

liggitt commented:
anything odd with GOROOT, GOPATH, etc?

derekwaynecarr commented:
/cc @derekwaynecarr tracking...

mikedanese committed:
f

vishh commented:
@mtaufen will you be able to pick this up from @mikedanese ?

mtaufen commented:
@vishh @mikedanese sure thing

mtaufen commented:
I will work on rebasing this today.

dchen1107 commented:
@mtaufen Thanks for picking up this work.

mtaufen committed:
Make it build!

Update external KubeletConfiguration type
Add defaults for new KubeletConfiguration fields
Modify some defaults to match upstream settings
Add/rename some conversion functions
Updated codegen
Fixed typos
Mike Danese caught that s.NodeLabels wasn't allocated, fix on line 118
of cmd/kubelet/app/options/options.go.
Provide list of valid sources in comment for HostNetworkSources field

mtaufen committed:
Fix unit test build errors

These tests can just use the empty string for the PodInfraContainerImage.

mtaufen commented:
Yay, the tests pass! Can @vishh and @timstclair please review this rebased PR :).

vishh commented on cmd/kubelet/app/options/options.go:
```diff
@@ -65,92 +44,20 @@ type KubeletServer struct {
-		},
+		AuthPath:             util.NewStringFlag("/var/lib/kubelet/kubernetes_auth"), // deprecated
+		KubeConfig:           util.NewStringFlag("/var/lib/kubelet/kubeconfig"),
+		SystemReserved:       make(utilconfig.ConfigurationMap),
```
Question: Why is system reserved and kube reserved not part of `kubeletconfiguration`?

mtaufen replied:
@mikedanese Can you comment on this?

mtaufen replied:
I talked to @mikedanese about this last week and we decided it makes sense to move `SystemReserved` and `KubeReserved` into `KubeletConfiguration`, I'll take care of this today.

vishh commented on pkg/apis/componentconfig/types.go:
```diff
@@ -158,13 +160,13 @@ type KubeletConfiguration struct {
 	AllowPrivileged bool `json:"allowPrivileged"`
 	// hostNetworkSources is a comma-separated list of sources from which the
 	// Kubelet allows pods to use of host network. Defaults to "*".
```
nit: We should specify the list of valid `sources`.

mtaufen replied:
Are the constants at `k8s.io/kubernetes/pkg/kubelet/types/pod_update.go` exhaustive or are there any others we should list?

mtaufen replied:
Done

vishh commented on pkg/apis/componentconfig/v1alpha1/defaults.go:
```diff
@@ -112,3 +132,193 @@ func SetDefaults_LeaderElectionConfiguration(obj *LeaderElectionConfiguration) {
 	}
 }
+
+func SetDefaults_KubeletConfiguration(obj *KubeletConfiguration) {
```
@mikedanese We currently have scripts that auto-generate documentation based on flag defaults. We should ideally update those scripts to use the defaults from here.

mtaufen replied:
What needs to be updated to make this change?

mikedanese replied:
I think this should happen automatically. We are just binding the flags to different addresses (fields in the componentconfig struct). The man page generation flow should be unaffected.

vishh commented on pkg/apis/componentconfig/v1alpha1/types.go:
```diff
@@ -139,3 +157,276 @@ type LeaderElectionConfiguration struct {
+	MaxPerPodContainerCount int64 `json:"maxPerPodContainerCount"`
+	// maxContainerCount is the maximum number of old instances of containers
+	// to retain globally. Each container takes up some disk space.
+	MaxContainerCount *int32 `json:"maxContainerCount"`
```
Why is this a pointer? Is it because is it is optional?

mtaufen replied:
I'll have to ask @mikedanese to double check, but generally speaking we use a pointer for numeric values where 0 is a valid option but a poor default (nil then implies default behavior). In this case, my guess would be that we allow a user to set 0, but in most scenarios you would not want to place that tight of a restriction on old container instances.

mikedanese replied:
Agree with @mtaufen. 0 is valid but not a good default.

timstclair replied:
We should really audit all these for optional values, and annotate them all with `omitempty` (see https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#optional-vs-required).

Any optional primitive field should really be a pointer, whether or not 0 is an acceptable default.

vishh commented on pkg/kubelet/dockertools/docker_test.go:
```diff
@@ -653,7 +652,7 @@ func TestFindContainersByPod(t *testing.T) {
 	np, _ := network.InitNetworkPlugin([]network.NetworkPlugin{}, "", nettest.NewFakeHost(nil), componentconfig.HairpinNone, "10.0.0.0/8")
 	// image back-off is set to nil, this test should not pull images
-	containerManager := NewFakeDockerManager(fakeClient, &record.FakeRecorder{}, nil, nil, &cadvisorapi.MachineInfo{}, options.GetDefaultPodInfraContainerImage(), 0, 0, "", &containertest.FakeOS{}, np, nil, nil, nil)
+	containerManager := NewFakeDockerManager(fakeClient, &record.FakeRecorder{}, nil, nil, &cadvisorapi.MachineInfo{}, "", 0, 0, "", &containertest.FakeOS{}, np, nil, nil, nil)
```
nit: Why set to `""` instead of the actual container name?

mtaufen replied:
The tests in this commit originally depended on some code in `k8s.io/kubernetes/cmd/kubelet/app/options/options.go`, and now that code has moved to `k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1/defaults.go`. @mikedanese and I discussed whether we should just make the tests depend on `defaults.go`, or reduce the maintenance burden by passing empty strings here. 

We determined that:
1. Changing the tests in this way didn't cause any of them to fail, so we aren't presently deriving coverage from having an accurate image name here. We therefore don't lose coverage with this change, but we do (very slightly) reduce the dependency tree, which is good. 
2. We will know if a future test requires an accurate image name, because such a test will fail without it. We can therefore drop the dependency without fear of accidentally missing coverage down the road. 

Given these conditions, we figured it makes the most sense to just pass `""` here.

vishh commented:
@mtaufen Can you address all the pending comments?

liggitt commented on cmd/kubelet/app/options/options.go:
```diff
@@ -65,92 +44,20 @@ type KubeletServer struct {
-			EvictionPressureTransitionPeriod: unversioned.Duration{Duration: 5 * time.Minute},
-			PodsPerCore:                      0,
-		},
+		AuthPath:             util.NewStringFlag("/var/lib/kubelet/kubernetes_auth"), // deprecated
```
why are these not represented as config struct fields?

mtaufen replied:
I'll move `SystemReserved` and `KubeReserved` into the config struct today. I talked to @mikedanese about these fields last week and IIRC the reason not to move `AuthPath` and `KubeConfig` into the config struct (yet) is that they are read from disk, but we do not yet have the option to read the kubelet config struct from disk. @mikedanese please correct me if I am not remembering correctly.

mikedanese replied:
AuthPath is required to connect to api-server. We should group that in the KubeConfig bucket and leave the ComponentConfig bucket for options that don't pertain to api-server connection.

mtaufen replied:
I'm currently rebasing again, it looks like the comment for `AuthPath` has updated to `Deprecated -- use KubeConfig instead`. So perhaps the functionality is already present in `KubeConfig` and we should we leave `AuthPath` where it is for legacy reasons?

mikedanese replied:
Yup. That sounds good.

mtaufen commented on pkg/apis/componentconfig/v1alpha1/defaults.go:
```diff
@@ -112,3 +132,193 @@ func SetDefaults_LeaderElectionConfiguration(obj *LeaderElectionConfiguration) {
+		obj.DockerExecHandlerName = "native"
+	}
+	if obj.DockerEndpoint == "" { // TODO: This isn't in NewKubeletServer in options.go in master...
+		obj.DockerEndpoint = "unix:///var/run/docker.sock"
```
In the `AddFlags` function in `k8s.io/kubernetes/cmd/kubelet/app/options/options.go`, `DockerEndpoint` is specified as `fs.StringVar(&s.DockerEndpoint, "docker-endpoint", s.DockerEndpoint, "If non-empty, use this for the docker endpoint to communicate with")`. The usage string seems to indicate that `""` is a valid default, so perhaps we shouldn't substitute `"unix:///var/run/docker.sock"`? @mikedanese @vishh

mikedanese replied:
"" is not valid. Defaulting it to "unix:///var/run/docker.sock" when you get a "" SGTM. Also DockerAPIEndpoint sounds fine. That struct was very literally translated from the cli options so you should change it to be more consistent.

mtaufen replied:
Ok, SGTM.

mtaufen committed:
Clean up API

Move SystemReserved and KubeReserved into KubeletConfiguration struct
Convert int64 to int32 for some external type fields so they match internal ones
tLS* to tls* for JSON field names
Fix dependency on removed options.AutoDetectCloudProvider
Change floats in KubeletConfiguration API to ints

mtaufen commented:
@vishh @mikedanese @timstclair please take a look :)

timstclair commented:
I think the API could really use some work: clean up names (frequency vs. period), clean up types (pointer, int32 vs 64, etc.), group related fields, etc.

IIRC the plan was to check this in as a direct translation of what we already had in Kubelet and then follow up with a cleanup of the API. Is that still the case?

mtaufen commented:
Whatever the original plan was, I'd be in favor of cleaning up the API in this PR. Better we get through API volatility while it's still in review vs after a merge.

timstclair commented:
Ok. In that case I'd suggest carefully reading through that [api-conventions](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md) doc, and applying it here. I also think it makes sense to group a lot of these fields in nested structs (mostly for readability), and I suspect some other issues will become clear while making those changes. There also might be some options that can be removed or deprecated.

luxas commented:
cc @kubernetes/sig-cluster-lifecycle

mtaufen commented:
I have a few final questions as I clean up the API:

I see inconsistent use of `float32` vs `float64` on `QPS` fields; is there any reason to prefer one over the other? It's a nit, but I like consistent type usage for fields with similar function. (fields are `RegistryPullQPS float64`, `EventRecordQPS float32`, and `KubeAPIQPS float32` internally and `RegistryPullQPS *float64`, `EventRecordQPS float32`, and `KubeAPIQPS float32` externally).

Also related to `QPS` fields, will the latter two ever have a nonzero default? They should probably be pointers on the external type if this is a possibility.

`MaxOpenFiles` is `int64`. How do I make sure that this is serialized and accepted as a string wrt API conventions for integer values > 53 bits? From API conventions:

> All numbers (e.g., uint32, int64) are converted to float64 by Javascript and some other languages, so any field which is expected to exceed that either in magnitude or in precision (specifically integer values > 53 bits) should be serialized and accepted as strings.

What is a sane default for `LockFilePath`? We aren't explicitly setting one at the moment, so it's presently set to `""` if no value is set. Is this ok?

I want to clarify my understanding of `omitempty`. As I understand it, we don't need it on the external type at all. On the internal type, my understanding is that you use it for non-pointer/non-built-in-nil typed values when you don't want the API server to complain about their omission from a request. Is this correct?

More reorganization can still be done in a separate PR after this one goes in, and the external type will be considered volatile for the time being, I just want to take care of these things while they're on my mind and then get this PR in so I can move forward with #27980.

timstclair commented:
>  see inconsistent use of float32 vs float64 on QPS fields; is there any reason to prefer one over the other? It's a nit, but I like consistent type usage for fields with similar function. (fields are RegistryPullQPS float64, EventRecordQPS float32, and KubeAPIQPS float32 internally and RegistryPullQPS *float64, EventRecordQPS float32, and KubeAPIQPS float32 externally).

I don't think those need to be floats? I can't imagine we'd ever want 0 < qps < 1, so I think an int32 would be [preferred](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#primitive-types)

> Also related to QPS fields, will the latter two ever have a nonzero default? They should probably be pointers on the external type if this is a possibility.

I'd say that when in doubt, go with a pointer.

> MaxOpenFiles is int64. How do I make sure that this is serialized and accepted as a string wrt API conventions for integer values > 53 bits?

I can't imagine MaxOpenFiles would every overflow 53 bits... (never say never WRT computers scaling... but I'd say not for a long time).

> What is a sane default for LockFilePath? We aren't explicitly setting one at the moment, so it's presently set to "" if no value is set. Is this ok?

Not sure...

> I want to clarify my understanding of omitempty. As I understand it, we don't need it on the external type at all. On the internal type, my understanding is that you use it for non-pointer/non-built-in-nil typed values when you don't want the API server to complain about their omission from a request. Is this correct?

I think (but I"m not sure) that you have external and internal backwards, but the simplest way to think of it is as a documentation bit. It should be safe to simply include it on any field that is conceptually optional. This might be a problem if the field has a valid zero-value, but then the fix is to make it a pointer.

timstclair commented on pkg/apis/componentconfig/v1alpha1/types.go:
```diff
@@ -415,7 +415,7 @@ type KubeletConfiguration struct {
 	EnableCustomMetrics bool `json:"enableCustomMetrics"`
 	// Comma-delimited list of hard eviction expressions.  For example, 'memory.available<300Mi'.
-	EvictionHard string `json:"evictionHard"`
+	EvictionHard *string `json:"evictionHard"`
```
I think there's no reason to have the internal & external versions differ for the time being. Also, this should have `omitempty`.

mtaufen replied:
@derekwaynecarr recently enabled eviction by default by setting the  `EvictionHard` threshold, so now `""` is no longer the correct default. But `""` is still a valid value, because, presumably, setting `""` is how you would disable eviction. So we need a pointer on the external. I can definitely add `omitempty` though.

timstclair commented:
Changes look good, but I think a lot of the fields might need to be pointers with omitempty eventually. Up to you whether you want to do that in this PR.

mikedanese commented:
omitempty is questionably useful here for a few reasons:

* omitempty only has effect in the serialization to json path, which unlike other apigroups, doesn't ever get called (at least in the initial design of this feature). We only ever read configuration.
* If we do start serializing these types to expose the componentconfig the main usecase that I can think of is for debugging a component. For this purpose, @smarterclayton pointed out that unlike other objects, it is probably useful to show the full effective configuration and not omit fields that have go's default values because these fields carry relevant information.

mtaufen commented:
Actually, one of the first things I was going to do after this PR was add a way to get the config from a Kubelet, to aid in debugging. So we probably will end up serializing these types. In this case, @smarterclayton's point holds.  

The api-conventions document seems to indicate that there is some API server logic that will reject POSTs or PUTs with unset fields not tagged `omitempty`. I suppose I misinterpreted this? (see the third bullet):

> Required fields have the opposite properties, namely:
> - They do not have an `omitempty` struct tag.
> - They are not a pointer type in the Go definition (e.g. bool otherFlag).
> - The API server should not allow POSTing or PUTing a resource with this field unset.

mtaufen commented:
On another note, `QPS` fields in other `componentconfig` types are `float*` as well. `KubeSchedulerConfiguration.KubeAPIQPS` is `float32`, for example. Might be worth tracking down if there was actually a reason for this.

mikedanese commented:
There's no obvious way to generate swagger doc for this since it's not hosted at an API endpoint.

mtaufen commented:
And this is because we are just reusing the endpoints for ConfigMaps, correct?

mikedanese commented:
Yes.

mtaufen commented:
Ok.

mtaufen commented:
@timstclair
> but I think a lot of the fields might need to be pointers with omitempty eventually. Up to you whether you want to do that in this PR.

Let's get this one in the PR that will group related fields into substructures.

mikedanese commented:
Also this is going into an alpha api so we have wiggle room.

mtaufen commented:
Right.

smarterclayton commented:
omitempty is kind of broken right now for maps and lists.  The third rule
you mention we don't enforce (or have any way to enforce).  So the second
point is definitely true, the first point is true for primitives, and for
arrays and maps we're about to do some fixing internally (which will lead
to omitempty not being empty).

We definitely recommend for optional arrays and maps that you define very
clearly the default behavior, because we've gotten it wrong a bunch.

On Tue, Jul 12, 2016 at 7:38 PM, Kubernetes Bot <notifications@github.com>
wrote:

> GCE e2e build/test *passed* for commit 2b1fdb6
> <https://github.com/kubernetes/kubernetes/commit/2b1fdb62530fa33d24991cefd17cb110e2b22d50>
> .
>
>    - Test Results
>    <https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/19837/kubernetes-pull-build-test-e2e-gce/48757>
>    - Build Log
>    <http://pr-test.k8s.io/19837/kubernetes-pull-build-test-e2e-gce/48757/build-log.txt>
>    - Test Artifacts
>    <https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/19837/kubernetes-pull-build-test-e2e-gce/48757/artifacts/>
>    - Internal Jenkins Results
>    <http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//48757>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/19837#issuecomment-232213719>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p3fsmyKMMhQIPO12eGO_1nbXP7Oxks5qVCV0gaJpZM4HIMNP>
> .
>

mtaufen commented:
Can you clarify what you mean by "which will lead to omitempty not being empty."

smarterclayton commented:
omitempty means "optional".  "optional" often has the implication that you
want to know whether the user specified a value or not.  Normal JSON
serialization without omitempty will distinguish between []string{} and
[]string(nil) (empty vs not specified).  When you add omitempty, nil is
always serialized as []string{} (empty) so there's no actual way to tell
whether the user sent you "empty" or didn't send anything at all.

On Wed, Jul 13, 2016 at 11:23 AM, Kubernetes Bot <notifications@github.com>
wrote:

> GCE e2e build/test *failed* for commit 288f4f9
> <https://github.com/kubernetes/kubernetes/commit/288f4f950dee7f0a62659644997640fedc12056d>
> .
>
>    - Test Results
>    <https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/19837/kubernetes-pull-build-test-e2e-gce/48871>
>    - Build Log
>    <http://pr-test.k8s.io/19837/kubernetes-pull-build-test-e2e-gce/48871/build-log.txt>
>    - Test Artifacts
>    <https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/19837/kubernetes-pull-build-test-e2e-gce/48871/artifacts/>
>    - Internal Jenkins Results
>    <http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//48871>
>
> Please reference the list of currently known flakes
> <https://github.com/kubernetes/kubernetes/issues?q=is:issue+label:kind/flake+is:open>
> when examining this failure. If you request a re-test, you must reference
> the issue describing the flake.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/19837#issuecomment-232390078>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p24gsbDtu20iMLCamih8RVQ9tLNrks5qVQLZgaJpZM4HIMNP>
> .
>

mtaufen commented:
So we'll soon have a way of making that distinction for omitempty fields as well? That's something I can look forward to :).

timstclair commented:
LGTM
